llrt_modules/src/modules/fs/file_handle.rs

```diff
@@ -0,0 +1,559 @@
+        let length = options_1
+            .length
+            .or(options_2.length)
+            .or(length.0)
+            .unwrap_or_else(|| buffer.len() - offset);
+
+        // It is not safe to pass the buffer from `ArrayBufferView` to `File::read`
+        // since the read is done in a different thread and we cannot garantee
+        // that multiple read calls are not done with the same buffer.
+        // Ideally, we should make our own version of `BufReader` to reuse the buffer
+        // instead of doing an allocation on each read.
+        let mut buf = vec![0u8; length];
+        let bytes_read = self
+            .file_mut(&ctx)?
+            .read(&mut buf)
+            .await
+            .or_throw_msg(&ctx, "Failed to read file")?;
+
+        let dst_buf = unsafe {
+            buffer
+                .as_bytes_mut()
+                .or_throw_msg(&ctx, "Buffer is detached")?
+        };
+        dst_buf[offset..].copy_from_slice(&buf);
+
+        let result = Object::new(ctx)?;
+        result.set("bytesRead", bytes_read)?;
+        result.set("buffer", buffer)?;
+        Ok(result)
+    }
+
+    async fn read_file<'js>(
+        &mut self,
+        ctx: Ctx<'js>,
+        options: Opt<Either<String, read_file::ReadFileOptions>>,
+    ) -> Result<Value<'js>> {
+        let size = self
+            .file(&ctx)?
+            .metadata()
+            .await
+            .map(|m| m.len() as usize)
+            .ok();
+        let mut bytes = Vec::new();
+        bytes
+            .try_reserve_exact(size.unwrap_or(0))
+            .or_throw_msg(&ctx, "Out of memory")?;
+
+        self.file_mut(&ctx)?
+            .read_to_end(&mut bytes)
+            .await
+            .or_throw_msg(&ctx, "Failed to read file")?;
+        read_file::handle_read_file_bytes(&ctx, options, bytes)
+    }
+
+    async fn stat(&self, ctx: Ctx<'_>) -> Result<Stat> {
+        let metadata = self
+            .file(&ctx)?
+            .metadata()
+            .await
+            .or_throw_msg(&ctx, "Can't stat file")?;
+        Ok(Stat::new(metadata))
+    }
+
+    async fn sync(&self, ctx: Ctx<'_>) -> Result<()> {
+        self.file(&ctx)?
+            .sync_all()
+            .await
+            .or_throw_msg(&ctx, "Can't sync file")
+    }
+
+    async fn truncate(&mut self, ctx: Ctx<'_>, len: Opt<u64>) -> Result<()> {
+        let len = len.0.unwrap_or(0);
+        self.file_mut(&ctx)?
+            .set_len(len)
+            .await
+            .or_throw_msg(&ctx, "Can't truncate file")
+    }
+
+    // Setting times not supported in tokio
+    // See https://github.com/tokio-rs/tokio/issues/6368
+    // async fn utimes(&mut self,  ctx: Ctx<'_>, atime: Value<'_>, mtime: Value<'_>) -> Result<()>
+
+    async fn write<'js>(
+        &mut self,
+        ctx: Ctx<'js>,
+        buffer_or_string: Either<ArrayBufferView<'js>, String>,
+        offset_or_options_or_position: Opt<Either<Either<usize, Null>, WriteOptions>>,
```

Comment:
So the way Either works is similar to an untagged serde enum. It will try each conversion and stop at the first that works.

There is a little runtime penalty since we need to clone Value for each conversion try which increments the reference counter but that is minimal.